### ISSUE TITLE
Gas Optimizations

### ISSUE THREAD

**Issue by code423n4** [code-423n4/2022-04-backd-findings#188]
### G-01: Unnecessary if statement in `processExpiredLocks`

[CvxCrvRewardsLocker.processExpiredLocks; L133-L145](https://github.com/code-423n4/2022-04-backd/blob/main/backd/contracts/CvxCrvRewardsLocker.sol#L133-L145)

Gas could be saved by changing:

```
    function processExpiredLocks(bool relock) external override returns (bool) {
        if (relock) {
            require(!prepareWithdrawal, Error.PREPARED_WITHDRAWAL);
        }

        if (relock) {
            ICvxLocker(CVX_LOCKER).processExpiredLocks(relock);
        } else {
            ICvxLocker(CVX_LOCKER).withdrawExpiredLocksTo(treasury);
        }

        return true;
    }
```

To:

```
    function processExpiredLocks(bool relock) external override returns (bool) {
        if (relock) {
            require(!prepareWithdrawal, Error.PREPARED_WITHDRAWAL);
            ICvxLocker(CVX_LOCKER).processExpiredLocks(relock);
        } else {
            ICvxLocker(CVX_LOCKER).withdrawExpiredLocksTo(treasury);
        }

        return true;
    }
```

### G-02: Changing parameters order could save some gas

[TopUpActionFeeHandler.payFees#L81-L109](https://github.com/code-423n4/2022-04-backd/blob/main/backd/contracts/actions/topup/TopUpActionFeeHandler.sol#L81-L109)

If `lpToken.safeTransferFrom` reverts `keeperAmount` and `treasuryAmount` will be unnecessarily calculated.

```
uint256 keeperAmount = amount.scaledMul(getKeeperFeeFraction());
uint256 treasuryAmount = amount.scaledMul(getTreasuryFeeFraction());
LpToken lpToken = LpToken(lpTokenAddress);

lpToken.safeTransferFrom(msg.sender, address(this), amount);
```

I suggest setting `keeperAmount` and `treasuryAmount` after `lpToken.safeTransferFrom` as seen below.

```

LpToken lpToken = LpToken(lpTokenAddress);

lpToken.safeTransferFrom(msg.sender, address(this), amount);

uint256 keeperAmount = amount.scaledMul(getKeeperFeeFraction());
uint256 treasuryAmount = amount.scaledMul(getTreasuryFeeFraction());
```

### Ga-03: For loop in `hasAnyRole` can be more gas efficient

[RoleManager.hasAnyRole#L73-L86](https://github.com/code-423n4/2022-04-backd/blob/main/backd/contracts/access/RoleManager.sol#L73-L86)

- It is unnecessary to initialize `i = 0` as zero is already the default value.
- Caching `roles.length` would save gas from calculating it every iteration
- Using a prefix `++i` is more effecient than using a postfix `i++`

In summary I suggest changing it from:

```
for (uint256 i = 0; i < roles.length; i++) {
    if (hasRole(roles[i], account)) {
        return true;
    }
}
```

To:

```
size = roles.length;
for (uint256 i; i < size; ++i) {
    if (hasRole(roles[i], account)) {
        return true;
    }
}
```

Also valid for(list not exhaustive):

- [CompoundHandler.\_getAccountBorrowsAndSupply#L135](https://github.com/code-423n4/2022-04-backd/blob/main/backd/contracts/actions/topup/handlers/CompoundHandler.sol#L135) -[TopUpKeeperHelper#L32](https://github.com/code-423n4/2022-04-backd/blob/main/backd/contracts/actions/topup/TopUpKeeperHelper.sol#L32)
  StakerVault - getstakedbyactions

**Comment by lazyherocat** [comment]
# #
# 1-Use assembly to write address storage values

**Context:**

[FraxlendPairDeployer.sol#L104-L107](https://github.com/code-423n4/2022-08-frax/blob/main/src/contracts/FraxlendPairDeployer.sol#L104-L107)
[FraxlendPairCore.sol#L171-L175](https://github.com/code-423n4/2022-08-frax/blob/main/src/contracts/FraxlendPairCore.sol#L171-L175)

**Proof Of Concept:**

The optimizer was turned on and set to 10000 runs

```js
contract GasTest is DSTest {
Contract0 c0;
Contract1 c1;

    function setUp() public {
        c0 = new Contract0();
        c1 = new Contract1();
    }
    
    function testGas() public {
        c0.admin(0x5B38Da6a701c568545dCfcB03FcB875f56beddC4);
        c1.admin(0x5B38Da6a701c568545dCfcB03FcB875f56beddC4);
    }
}


contract Contract0 {
    
    address public foo;
    
    function admin(address _implementation) external {
        foo = _implementation;
    }
}

contract Contract1 {
    
    address public foo;
    
    function admin(address _implementation) external {
        assembly {
        sstore(foo.slot, _implementation)
        }
    }
}
```

**Gas Report:**

```js
╭───────────────────────────────────────────┬─────────────────┬───────┬────────┬───────┬─────────╮
│ src/test/GasTest.t.sol:Contract0 contract ┆                 ┆       ┆        ┆       ┆         │
╞═══════════════════════════════════════════╪═════════════════╪═══════╪════════╪═══════╪═════════╡
│ Deployment Cost                           ┆ Deployment Size ┆       ┆        ┆       ┆         │
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┤
│ 66917                                     ┆ 366             ┆       ┆        ┆       ┆         │
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┤
│ Function Name                             ┆ min             ┆ avg   ┆ median ┆ max   ┆ # calls │
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┤
│ admin                                     ┆ 22363           ┆ 22363 ┆ 22363  ┆ 22363 ┆ 1       │
╰───────────────────────────────────────────┴─────────────────┴───────┴────────┴───────┴─────────╯
╭───────────────────────────────────────────┬─────────────────┬───────┬────────┬───────┬─────────╮
│ src/test/GasTest.t.sol:Contract1 contract ┆                 ┆       ┆        ┆       ┆         │
╞═══════════════════════════════════════════╪═════════════════╪═══════╪════════╪═══════╪═════════╡
│ Deployment Cost                           ┆ Deployment Size ┆       ┆        ┆       ┆         │
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┤
│ 51705                                     ┆ 290             ┆       ┆        ┆       ┆         │
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┤
│ Function Name                             ┆ min             ┆ avg   ┆ median ┆ max   ┆ # calls │
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┤
│ admin                                     ┆ 22324           ┆ 22324 ┆ 22324  ┆ 22324 ┆ 1       │
╰───────────────────────────────────────────┴─────────────────┴───────┴────────┴───────┴─────────╯
```
